pkg/cmd/experimental/buildchain/buildchain.go

```diff
@@ -0,0 +1,367 @@
+	cmd.Flags().StringP("output", "o", "json", "Output format of dependency tree(s)")
+	return cmd
+}
+
+// getRepos iterates over a given set of build configurations
+// and extracts all the image repositories on which other
+// image repositories are based
+func getRepos(configs []buildapi.BuildConfig) map[string][]string {
+	avoidDuplicates := make(map[string][]string)
+	for _, cfg := range configs {
+		for _, tr := range cfg.Triggers {
+			if tr.ImageChange != nil && tr.ImageChange.From.Name != "" {
+				var repo string
+				switch tr.ImageChange.From.Namespace {
+				case "":
+					repo = join(cfg.Namespace, tr.ImageChange.From.Name)
+				default:
+					repo = join(tr.ImageChange.From.Namespace, tr.ImageChange.From.Name)
+				}
+
+				var tag string
+				switch tr.ImageChange.Tag {
+				case "":
+					tag = "latest"
+				default:
+					tag = tr.ImageChange.Tag
+				}
+
+				uniqueTag := true
+				for _, prev := range avoidDuplicates[repo] {
+					if prev == tag {
+						uniqueTag = false
+						break
+					}
+				}
+				if uniqueTag {
+					avoidDuplicates[repo] = append(avoidDuplicates[repo], tag)
+				}
+			}
+		}
+	}
+
+	return avoidDuplicates
+}
+
+// findRepoDeps accepts an image repository and a list of build
+// configurations and returns the dependency tree of the specified
+// image repository
+func findRepoDeps(repo, tag string, all bool, buildConfigList []buildapi.BuildConfig) (*ImageRepo, error) {
+	root := &ImageRepo{
+		FullName: repo,
+		Tags:     []string{tag},
+	}
+
+	namespace, name, err := split(repo)
+	if err != nil {
+		return nil, err
+	}
+
+	// Search all build configurations in order to find the image
+	// repositories depending on the specified image repository
+	var childNamespace, childName, childTag string
+	for _, cfg := range buildConfigList {
+		for _, tr := range cfg.Triggers {
+			// Setup zeroed fields to their default values
+			if tr.ImageChange != nil && tr.ImageChange.From.Namespace == "" {
+				tr.ImageChange.From.Namespace = cfg.Namespace
+			}
+			if tr.ImageChange != nil && tr.ImageChange.Tag == "" {
+				tr.ImageChange.Tag = "latest"
+			}
+
+			if tr.ImageChange != nil &&
+				namespace == tr.ImageChange.From.Namespace &&
+				name == tr.ImageChange.From.Name &&
+				tag == tr.ImageChange.Tag {
+
+				// Either To & Tag or DockerImageReference will be used as output
+				if cfg.Parameters.Output.To != nil && cfg.Parameters.Output.To.Name != "" {
+					childName = cfg.Parameters.Output.To.Name
+					childTag = cfg.Parameters.Output.Tag
+					if cfg.Parameters.Output.To.Namespace != "" {
+						childNamespace = cfg.Parameters.Output.To.Namespace
+					} else {
+						childNamespace = cfg.Namespace
+					}
+				} else {
+					ref, err := imageapi.ParseDockerImageReference(cfg.Parameters.Output.DockerImageReference)
+					if err != nil {
+						return nil, err
+					}
+					childName = ref.Name
+					childTag = ref.Tag
+					childNamespace = cfg.Namespace
```

Comment:
i don't think we need to change this, but it's worth noting that the imagereference's namespace (eg "openshift" in a pullspec like openshift/ruby-20-centos7:foo) could actually be the namespace of the imagerepo.  unfortunately it's impossible to know, so using the buildconfig's namespace is just as reasonable.

fundamentally if your buildconfig outputs to an imageref instead of an imagerepo, things get pretty fuzzy.